### PR TITLE
Add/user menu component

### PR DESCRIPTION
**Pull request by kdevnel**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes https://github.com/Automattic/woocommerce.com/issues/17509

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Before checking out the branch, checkout `trunk` and set up a connection to WooCommerce.com with a test account. (you can do this by clicking "Connect" at `/wp-admin/admin.php?page=wc-addons&section=helper`).
2. Checkout this branch
    a. from the `plugins/woocommerce` folder, you may need to run `composer dump-autoload`
    b. from the `plugins/woocommerce-admin` folder, you may need to run `pnpm install` if you get dependency issues
3. Build the project by going to `plugins/woocommerce` and running `pnpm run --filter=woocommerce build`
4. Visit the Extensions page (WP Env URL is http://localhost:8888/wp-admin/admin.php?page=wc-admin&path=%2Fextensions)
5. The component should appear as a small icon above the Discover and Extensions tabs.
6. Check against the MVP design in Figma - you should see the **connected** version, and the **local** user avatar.
7. Test the logged out version by editing [this line](https://github.com/woocommerce/woocommerce/blob/6557db1a5abc3f20d9d539146c0706362e14dfb3/plugins/woocommerce/includes/admin/helper/class-wc-helper-admin.php#L24) to `'isConnected' => false,`
    - **note no rebuild is required for this test**

### Screenshots
**Logged out:**
Screenshot 2023-07-26 at 16 56 00

**Logged in:**
Screenshot 2023-07-26 at 16 56 17

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-admin.php
Comment (kdevnel) [inline plugins/woocommerce/includes/admin/helper/class-wc-helper-admin.php]: Note to reviewers: We will need to return to review this URL still works later once this issue is resolved: https://github.com/orgs/Automattic/projects/366/views/43?pane=issue&itemId=32504577

File: plugins/woocommerce-admin/client/marketplace/components/header-account/header-account.scss
Comment (andfinally) [inline plugins/woocommerce-admin/client/marketplace/components/header-account/header-account.scss]: The linter doesn't seem to complain when we don't order rules alphabetically, like it does on WCCOM... but it would still be nice to do it, IMHO. Ignore if this is too pedantic!

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-admin.php
Comment (andfinally) [inline plugins/woocommerce/includes/admin/helper/class-wc-helper-admin.php:15]: It would be useful to add a line here about what the helper is for.

```suggestion
 * The main entry-point for all things related to the Helper.
 * The Helper manages the connection between the store and 
 * an account on WooCommerce.com.
```

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-admin.php
Comment (andfinally) [inline plugins/woocommerce/includes/admin/helper/class-wc-helper-admin.php:37]:
```suggestion

	public static function add_marketplace_settings( $settings ) {
```

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-admin.php
Comment (andfinally) [inline plugins/woocommerce/includes/admin/helper/class-wc-helper-admin.php:75]:
```suggestion

		$connect_url = add_query_arg(
```

File: plugins/woocommerce-admin/client/marketplace/components/header-account/header-account.tsx
Comment (andfinally) [inline plugins/woocommerce-admin/client/marketplace/components/header-account/header-account.tsx]: I wonder if we're getting this from the right place? This is returning a generic grey avatar for me. But the old skool PHP connect panel is [calling](https://github.com/woocommerce/woocommerce/blob/ef015f740f89c35cebf34bc0c39a971be6723b68/plugins/woocommerce/includes/admin/helper/views/html-section-account.php#L9) `get_avatar( $auth_user_data['email'], 48 )`, which shows my usual avatar.

image

**Review by andfinally — CHANGES_REQUESTED** [review]
Very nice, love it! Just a few points, some of which are design points really. (cc @poligilad-auto).

The markup for the `WooCommerce Account` menu item has an `a` tag within a `button` – that shouldn't happen! It means that when you press your tab key to navigate between the links in the menu, the focus goes first on the button, then on the link inside it. The tabbing navigation should move smoothly from one item to the next.

### Design points

Personally I don't think the `WooCommerce.com Account` link makes sense when the person isn't connected – I would suggest we hide it.

It's also odd that the person's email is clickable, but doesn't do anything. I would vote for disabling that option, till it has a function.

_Edit -- Deleted comments based on the i4 design._

**Comment by poligilad-auto** [comment]
Hey, I didn't checkout the live version yet.

Just wanted to point out that the latest version is the [MVP page](https://www.figma.com/file/2sXEw33oLyTZ7Mu3808sed/In-App-Marketplace?type=design&node-id=1018%3A176413&mode=design&t=5v0KZ6lDvXe1fEr1-1), this is the latest design 👇 @andfinally 

image

> Personally I don't think the WooCommerce.com Account link makes sense when the person isn't connected – I would suggest we hide it.

Maybe we can rename it to `WooCommerce.com` when not connected? I think it would be useful to still have it there.

> It's also odd that the person's email is clickable, but doesn't do anything

I think this can link to https://woocommerce.com/my-dashboard/. WDYT?

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-admin.php
Comment (andfinally) [inline plugins/woocommerce/includes/admin/helper/class-wc-helper-admin.php]: Does this name need to be camel case? Snake case is usual for PHP in WP.

File: plugins/woocommerce-admin/client/marketplace/components/header-account/header-account.scss
Comment (kdevnel) [reply plugins/woocommerce-admin/client/marketplace/components/header-account/header-account.scss]: Agreed, updating in ebcb029b2e0561e085c47546b8d68583ea092824

File: plugins/woocommerce-admin/client/marketplace/components/header-account/header-account.tsx
Comment (kdevnel) [reply plugins/woocommerce-admin/client/marketplace/components/header-account/header-account.tsx]: Ah! That function pulls the gravatar directly. I was pulling an existing bit of data that is based on the admin email address. In our case, that email is the local environment email which doesn't have a gravatar! Fixed this to now pull that directly using the WCCOM account email address.

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-admin.php
Comment (kdevnel) [reply plugins/woocommerce/includes/admin/helper/class-wc-helper-admin.php:15]: Agreed!

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-admin.php
Comment (kdevnel) [reply plugins/woocommerce/includes/admin/helper/class-wc-helper-admin.php]: Oooh, good spot. Too much jumping around between PHP and JS!

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-admin.php
Comment (kdevnel) [reply plugins/woocommerce/includes/admin/helper/class-wc-helper-admin.php:75]: Done.

**Comment by poligilad-auto** [comment]
Nice work @kdevnel!

Must say that I'm really impressed by the exact implementation here, with correct spacing and specs for almost everything 👌

**Some comments:**

I hope we are using variables for all the colors? It seems like some of the black and gray colors aren't using variables.
I'm not sure where we have those setup. If we are, could you please share the file with me so I can bookmark it?

1. The header: the top part currently has [10px padding from the top](https://prnt.sc/t8pu28emtUyC), which makes the header too high. Can we remove this spacing and instead apply this spacing inside the top part of the header, so the contents won't be centered but "shifted down 10px?"
2. The title "Extensions" should be 16px.
3. Search bar:
- The placeholder text inside the search bar should say "Search for extensions".
- The [component](https://wordpress.github.io/gutenberg/?path=/story/components-searchcontrol--default) isn't behaving as expected. It should have a _white fill_ when focused with a _blue border_, plus the _search icon_ should change to a _close icon_.
4. Tab navigation: 
- Change "Extensions" to "Browse".
- Remove the left and right padding of 32px for the whole tab section.
- Remove `margin: 0 1.5rem 0 0` from the individual tabs. This causes too much spacing in between, it should only be 24px.
5. Main content: 
- `.woocommerce-marketplace__content` change top padding to 96px. I agreed this will be the spacing for new pages (it should be 48px below the header, and this is how to achieve it with the current padding).
- The h1 titles should be 20px.
6. Footer USPs: 
- Change the bottom Woo Marketplace to `Gray 50 (#646970)`.
- Change to "30-day money-back guarantee".
- Change to "Safe and secure online payment".
- On mobile: change the right & left padding for the footer content to 16px. (This is my bad, I didn't update it in the designs)
7. User area:
- Change "WooCommerce.com Account" to "WooCommerce.com account". We are using `Sentence case` for everything.

Regarding the copy, it might slightly change as I submitted the MVP for [a copy review](https://woomarketingstudio.wordpress.com/2023/07/24/studio-request-ux-copy-in-app-marketplace-mvp/#comment-13989) from the creative studio. We will hear back from them sometime next week.

**Comment by poligilad-auto** [comment]
Regarding the links from the user menu, I'm a fan of opening a new tab, in order to not loose where you were with WP-admin. WDYT @kdevnel?

**Comment by kdevnel** [comment]
Thanks for the awesome feedback @poligilad-auto ! I've created a [separate issue to handle your feedback](https://github.com/orgs/Automattic/projects/366/views/43?pane=issue&itemId=34472763) on items that aren't the user menu. We will probably be best batching it all together and handling it once all components are merged together on the feature branch.

I have implemented the sentence case changes you suggested and will push that version up soon. I haven't done anything about the opening in a new tab for links though. It might be difficult if not impossible with the component we're using right now without a lot of time. Either way, I am also a fan of not enforcing browser behaviour for various reasons. That said, If it's important enough, we can always revisit it at a later stage for sure!

**Comment by Dan-Q** [comment]
> Either way, I am also a fan of not enforcing browser behaviour for various reasons.

This. I haven't seen any recent resarch, but studies always _used_ to show that (a) experienced users dislike forced-open-in-new-tab because if they wanted something opened in a new tab they'd have opened it in a new tab, and (b) inexperienced users dislike open-in-new-tab because it breaks the way they expect the "back" button to work.

(b) might be less-relevant now because the "back" button has fallen out of fashion (mostly because badly-written websites and or dark patterns) have taught users that they can't rely on it any more. Also, more web users are "experienced" than once! But (a) still stands.

File: plugins/woocommerce-admin/client/marketplace/components/header-account/header-account.scss
Comment (andfinally) [inline plugins/woocommerce-admin/client/marketplace/components/header-account/header-account.scss:52]: Do we need to nest these?

File: plugins/woocommerce-admin/client/marketplace/components/header-account/header-account-modal.tsx
Comment (andfinally) [inline plugins/woocommerce-admin/client/marketplace/components/header-account/header-account-modal.tsx:11]:
```suggestion
```

File: plugins/woocommerce-admin/client/marketplace/components/header-account/header-account-modal.tsx
Comment (andfinally) [inline plugins/woocommerce-admin/client/marketplace/components/header-account/header-account-modal.tsx:40]: Woah, now I'm sorry I pushed for these BEM classnames! In future we can rethink what we're doing – treating the whole page as the block in BEM leads to this kind of name.

**Review by andfinally — COMMENTED** [review]
Lovely, thanks for the tweaks! 💯 

Something weird happens to the modal on mobile.



When I open the menu, the first item shows as selected. It's not really what I'd expect.

image

I also think it's weird that we link both the email and the `WooCommerce.com account` item to the same URL. But tis a minor quibble we can think about later on.

There are some [linting](https://github.com/woocommerce/woocommerce/actions/runs/5692727610/job/15430327289?pr=39383) errors from `class-wc-helper-admin.php`. Maybe we can get rid of the [template change warning](https://github.com/woocommerce/woocommerce/actions/runs/5692727612/job/15430319121?pr=39383) by copying that file from trunk, so there are no changes?

File: plugins/woocommerce-admin/client/marketplace/components/header-account/header-account.scss
Comment (kdevnel) [reply plugins/woocommerce-admin/client/marketplace/components/header-account/header-account.scss:52]: Sadly, in this case, I needed the specificity to override the component's built-in styling. Happy to break it out, but I'd still need to keep the specificity to do it. I have no strong opinions about how to handle that, so happy to go either way!

File: plugins/woocommerce-admin/client/marketplace/components/header-account/header-account-modal.tsx
Comment (kdevnel) [reply plugins/woocommerce-admin/client/marketplace/components/header-account/header-account-modal.tsx:11]: In this case, we have no internal dependencies, but I think we get liniting errors without the comment block. The space is to indicate that the export on line 12 is not associated with the comment.

File: plugins/woocommerce-admin/client/marketplace/components/header-account/header-account-modal.tsx
Comment (kdevnel) [reply plugins/woocommerce-admin/client/marketplace/components/header-account/header-account-modal.tsx:40]: Haha, I'm sure I could have shortened these. It's not ideal, but this is likely to be the most specific we need to get, so I don't think we will have anything longer at least.

**Comment by kdevnel** [comment]
@andfinally I've made some changes ready to review.

> Something weird happens to the modal on mobile.

Fixed! It now works the same as the component in Storybook. I had overridden a style that broke the mobile behaviour.

> When I open the menu, the first item shows as selected. It's not really what I'd expect.

Good point, although I don't have a good solution. This is the keyboard focus for accessibility and is the default behaviour. I see no way to disable it without breaking accessibility 🤔 

> I also think it's weird that we link both the email and the WooCommerce.com account item to the same URL. But tis a minor quibble we can think about later on.

Agreed, I'm not sure the best solution here, but happy to change this as necessary.

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-admin.php
Comment (andfinally) [inline plugins/woocommerce/includes/admin/helper/class-wc-helper-admin.php]: Maybe I'm being paranoid, but I'm wondering about the situation where `WC_Helper_Options::get( 'auth_user_data' )` isn't undefined, but doesn't have `['email']`. Perhaps we should try and get user data first, then check if it has this value.

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-admin.php
Comment (andfinally) [inline plugins/woocommerce/includes/admin/helper/class-wc-helper-admin.php:48]: I would have a line break before the return statement, like we do in `get_connection_url`. The WordPress code style got me into the habit of doing this.

```suggestion

		return $settings;
```

File: plugins/woocommerce-admin/client/marketplace/components/header-account/header-account.scss
Comment (andfinally) [reply plugins/woocommerce-admin/client/marketplace/components/header-account/header-account.scss:52]: Cool, just checking!

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-admin.php
Comment (andfinally) [inline plugins/woocommerce/includes/admin/helper/class-wc-helper-admin.php:71]: This will hopefully silence the [PHPCS warning](https://github.com/woocommerce/woocommerce/actions/runs/5715945454/job/15486367120?pr=39383) for the moment.

```suggestion
			// phpcs:ignore Generic.Commenting.Todo.TaskFound
			// TODO: this may be needed to make the URL work.
			// include WC_Helper::get_view_filename( 'html-oauth-start.php' ).
```

**Review by andfinally — COMMENTED** [review]
Looks great, thanks for the tweaks! I just had one quibble about the `WC_Helper_Options::get( 'auth_user_data' )` array.

About the selection of the first item in the dropdown... I see – so it's the `DropdownMenu` component that is doing this. I see the same thing in the Display menu on the Woo homepage, and other places.

image

Strangely, I don't see it in the product block editor three dots menu. But I can't see any code differences that might explain it. (I tried adding `isSelected={ false }` to the offending menu item, but it didn't help.)

image

In that case it seems OK to go with it.

**Comment by jonathansadowski** [comment]
It looks like Atlas was assigned as a reviewer via Codeowners because an [earlier push](https://github.com/woocommerce/woocommerce/compare/a834db53ffa0d39835737c37bcb27ecaabdf926a..ea0def80f8bd933d9cc7959d34e306fdce3f8736) to this PR inadvertently touched a file in the `.github` directory. That change no longer exists in this PR, so I'm going to resign as a reviewer, but if there's anything you do need our help reviewing, please let us know.

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-admin.php
Comment (kdevnel) [reply plugins/woocommerce/includes/admin/helper/class-wc-helper-admin.php:71]: Happy to do this, but do we want to silence this? This should be resolved fully before we finalise the project and so the warning is a reminder. It could be a little easy to miss the comment by the time we merge. 🤔

Comment (andfinally) [reply plugins/woocommerce/includes/admin/helper/class-wc-helper-admin.php:71]: Ya, that was at the back of my mind when I suggested this. We could merge the PR without it.

**Comment by andfinally** [comment]
I forgot a small CSS thing – when the menu icon is active (or focused), the outline doesn't match the size of the icon.

image

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-admin.php
Comment (andfinally) [inline plugins/woocommerce/includes/admin/helper/class-wc-helper-admin.php]: It would be nice to specify a default value, I reckon.

```suggestion
		$auth_user_data  = WC_Helper_Options::get( 'auth_user_data', array() );
```

**Review by andfinally — APPROVED** [review]
Thanks for the further tweaks! Looking fab. As you pointed, out the issue with the focus outline was on the feature branch. 🤦‍♂️

File: plugins/woocommerce-admin/client/marketplace/components/header-account/header-account-modal.tsx
Comment (raicem) [inline plugins/woocommerce-admin/client/marketplace/components/header-account/header-account-modal.tsx:23]: I'd probably consider putting them under a `useEffect` to make sure they only run once. Not a huge blocker though.

Edit: Well, I mistakenly thought these are function calls. I now realize they are just function definitions. Sorry!

File: plugins/woocommerce-admin/client/marketplace/components/header-account/header-account-modal.tsx
Comment (raicem) [inline plugins/woocommerce-admin/client/marketplace/components/header-account/header-account-modal.tsx:32]: JFYI, I am going for the prefix `woo-marketplace` in my branch. In case you want to update here!

**Review by raicem — APPROVED** [review]
I think this looks and works perfectly. Left some non-blocker comments.

File: plugins/woocommerce-admin/client/marketplace/components/header-account/header-account-modal.tsx
Comment (kdevnel) [reply plugins/woocommerce-admin/client/marketplace/components/header-account/header-account-modal.tsx:32]: Nice. I may end up doing that later once all the components are merged. The styles share a lot of styles with other components that use `woocommerce-marketplace__` and doing it here may actually cause more confusion and require editing beyond this component at this stage.

**Pull request merged**
